CHANGELOG.adoc

```diff
@@ -5,6 +5,7 @@
 === Architectural decision records
 
 - [ADR-069] Add support for children layout strategy
+- [ADR-070] Add the possibility to hide or fade a graphic element
```

Comment:
It could be better to by a tiny bit more precise since we have countless of "stuff" which can be considered as "graphical elements" with a title like "Add support to hide or fade diagram's views"

@pcdavid In Sirius Desktop all the graphical stuff in a diagram is called "view" right? like nodes, edges, decorators etc?

Reply:
We use the term, and variables which expose these elements are indeed called `view*`, but strictly speaking there is no type called "View" (there is `DView` but it's the top-level element of an aird).
In Sirius Desktop we have:
* `DDiagramElement extends DRepresentationElement extends DSemanticDecorator` as the most general common type for all elements on a diagram.
* Then `DNode`, `DNodeContainer` and `DEdge` are subtypes of `DDiagramElement`

`DDiagramElement` is the one which has a `boolean isVisible()` attribute, but this one is "derived" (not in the common EMF sense, but via a pre-commit listener, `VisibilityUpdater`) from another attribute: `EList<GraphicalFilter> getGraphicalFilters()`.

Each diagram element can have one or more `GraphicalFilter` attached, which are used to compute the element's visibility (as in `isVisible`), and importantly, "adjacent" elements visiblity too. For example an edge is visible only if both its source *and* target are visible (see https://github.com/pcdavid/org.eclipse.sirius/blob/master/plugins/org.eclipse.sirius.diagram/src/org/eclipse/sirius/diagram/business/internal/helper/display/DisplayServiceImpl.java#L221); a label can be hidden either because it was hidden itself explicitly or because its parent node is invisible, etc.

`GraphicalFilter`s act like "modifiers". The main ones are:
* `HideFilter`: "basic" hide/reveal feature as expected in this ADR;
* `FoldingFilter`: can be applied to some edges (think an edge from a parent node to its children) to "fold" the edge and hide the whole sub-tree of elements "below" the edge;
* `CollapseFilter`: can be applied to a node or container to reduce its size and hide all its content.

This may seem over-complicated, but we need to store and remember each of these flags to restore the proper state when an element is revealed.

Basically an element can become "invisible" on the diagram for different, cumulative reasons, and removing *one* of these reasons is not enough to make it visible again, so a `boolean` is not enough.

Reply:
Do I use `diagram elements` or `diagram's views`?
In this ADR, I propose to use an enum type to describe the state of the diagram element. This can be compared to a boolean because different possible states are incompatible with each other. Thus, an element is invisible (or faded) or not regardless of the reason. Do we have to implement a more flexible architecture to be able to handle the type of situation you mentioned in the future or should we keep it simple as I propose for the moment?

Reply:
I think that there are two types of information we need to have for two different needs:

- On the frontend, we need to know if an element should be displayed on not. For that, we can use several solutions such as "not returning the element in question from the backend", "use a boolean to indicate that its hidden" or even "use an enum to indicate that its hidden"
- On the backend, we need to know why an element should be displayed. For that, we will need the details involved for example "some edge is hidden because it is between two nodes which are in two hidden containers because they are collapsed" (even if collapsing is out of scope for now)

So we may need to keep the modifiers placed on nodes and edges like in Sirius Desktop `Node#modifiers: List<Modifier>` with :

```
enum Modifier {
  HIDDEN,
  FADED,
  (maybe COLLAPSED one day)
}
```

But we may also need, for the frontend, to compute (and store?) the impact of those modifiers. This would prevent us from having to compute this impact every time we need to process a diagram. This information could be something like that: `Node#state: Modifier` and `Edge#state: Modifier`.

For example, let's consider the diagram in this screenshot: 

<img width="1366" alt="Capture d’écran 2022-10-14 à 15 26 13" src="https://user-images.githubusercontent.com/422587/195858318-7bdc5c81-3579-47b1-bb83-628633339658.png">

If `Central_Unit` has for `modifiers` the values `[HIDDEN, FADED]`, then the following elements would be impacted with the following `state`:

- `DSP`: `HIDDEN`
- `Motion_Engine`: `HIDDEN`
- The fan with the label `100`: `HIDDEN`
- The edge with the label `10` between `DSP` and `Motion_Engine`: `HIDDEN`
- The edge with the label `6` between `Radar_Capture` and `Motion_Engine`: `HIDDEN`
- The edge with the label `10` between `GPU` and `Motion_Engine` should have the state `HIDDEN`

If on top of that, the `GPU` has for `modifiers` the values `[HIDDEN, FADED]`, then:

- The edge with the label `6` between `Back_Camera` and `GPU` should also have the state `HIDDEN`

As such, we need two pieces of information, the `modifiers` applied on the elements by the user and their impact on the `state` of a diagram element.

What do you think? @theogiraudet @pcdavid

Reply:
That sounds good to me, I don't see any case where this proposal doesn't work.

I see that you use a `List` to store the different modifiers. Is there a situation where we can have the same modifier more than once, e.g. `[HIDDEN, HIDDEN]` or is this an impossible case?

Reply:
I don't think we have any reason to have a `List` instead of a `Set` indeed